README.md
# Flutter Horizontal Featured List

A Flutter package to create Horizontal Featured List. It helps you in saving time during the development of many same designs. It will be useful for your awesome app.

![Demo](./demo.gif)

### Example

```dart
import 'dart:math' as math;
import 'package:flutter_horizontal_featured_list/flutter_horizontal_featured_list.dart';

class _MyAppState extends State<MyApp> {
  List<Color> colors = List.generate(20, (index) => randomColor());

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Example'),
        ),
        backgroundColor: Colors.white,
        body: Center(
          child: HorizontalFeaturedList(
            itemColorBuilder: (context, index) => colors[index],
            itemCount: colors.length,
            itemBuilder: (BuildContext context, int index) {
              return Column(
                children: [
                  Row(
                    children: <Widget>[
                      Icon(
                        Icons.face,
                        size: 24,
                        color: Colors.black45,
                      ),
                      Expanded(
                        child: Padding(
                          padding: const EdgeInsets.only(left: 8.0),
                          child: Text(
                            '($index) What is Lorem Ipsum?',
                            style: TextStyle(
                              fontSize: 16,
                              fontWeight: FontWeight.w500,
                            ),
                          ),
                        ),
                      ),
                    ],
                  ),
                  Expanded(
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Text(
                        '($index) Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
                        style: TextStyle(
                          fontSize: 12,
                        ),
                      ),
                    ),
                  ),
                ],
              );
            },
            onPressedItem: () {},
            onPressedSeeAll: () {},
            titleText: 'Ongoing Projects',
            seeAllText: 'See All',
          ),
        ),
      ),
    );
  }
}

Color randomColor() {
  var g = math.Random.secure().nextInt(255);
  var b = math.Random.secure().nextInt(255);
  var r = math.Random.secure().nextInt(255);
  return Color.fromARGB(255, r, g, b);
}
```

### Development environment

```
[✓] Flutter (Channel stable, 2.0.5, on macOS 11.2.3 20D91 darwin-x64, locale en-VN)
    • Flutter version 2.0.5
    • Framework revision adc687823a (11 days ago), 2021-04-16 09:40:20 -0700
    • Engine revision b09f014e96
    • Dart version 2.12.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode_12.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
```

CHANGELOG.md
## [1.0.1+1] - 25/06/2021

* Fix demo gif.

## [1.0.1] - 11/06/2021

* First release version.


lib/flutter_horizontal_featured_list.dart
import 'dart:math';

import 'package:flutter/material.dart';

/// {@tool snippet}
/// This is a sample of a [HorizontalFeaturedList] widget.
///
/// ```dart
///           HorizontalFeaturedList(
///             itemColorBuilder: (context, index) => colors[index],
///             itemCount: colors.length,
///             itemBuilder: (BuildContext context, int index) {
///               return Column(
///                 children: [
///                   Row(
///                     children: <Widget>[
///                       Icon(
///                         Icons.face,
///                         size: 24,
///                         color: Colors.black45,
///                       ),
///                       Expanded(
///                         child: Padding(
///                           padding: const EdgeInsets.only(left: 8.0),
///                           child: Text(
///                             '($index) What is Lorem Ipsum?',
///                             style: TextStyle(
///                               fontSize: 16,
///                               fontWeight: FontWeight.w500,
///                             ),
///                           ),
///                         ),
///                       ),
///                     ],
///                   ),
///                   Expanded(
///                     child: Padding(
///                       padding: const EdgeInsets.all(8.0),
///                       child: Text(
///                         '($index) Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
///                         style: TextStyle(
///                           fontSize: 12,
///                         ),
///                       ),
///                     ),
///                   ),
///                 ],
///               );
///             },
///             onPressedItem: () {},
///             onPressedSeeAll: () {},
///             titleText: 'Ongoing Projects',
///             seeAllText: 'See All',
///           )
/// ```
/// {@end-tool}
class HorizontalFeaturedList extends StatefulWidget {
  /// lets the [ListView] knows its children size, caused by it is placed inside a [Column].
  final double itemHeight;

  /// is the [TextStyle] use for <Title>.
  final TextStyle titleTextStyle;

  /// is the [TextStyle] use for <See All>.
  final TextStyle seeAllTextStyle;

  /// <See All> action.
  final Function() onPressedSeeAll;

  /// is the spacing from parent to the left side of <Title>, and from parent to the right side of <See All>, and spacing between <Featured List View> items also.
  final double margin;

  /// is the padding inside each of ListView items.
  final EdgeInsets itemPadding;

  /// is used for pressing animation on ListView items.
  final Color overlayColor;

  /// is the ListView item border radius.
  final BorderRadius itemBorderRadius;

  /// action to navigate to a detail screen.
  final Function() onPressedItem;

  /// See also [ListView.builder]
  final Widget Function(BuildContext context, int index) itemBuilder;

  /// See also [ListView.builder]
  final int itemCount;

  /// build the ListView item background color if needed.
  final Color Function(BuildContext context, int index)? itemColorBuilder;

  /// <Title> text.
  final String titleText;

  /// <See All> button text.
  final String seeAllText;

  /// lets the horizontal [ListView] knows its children size.
  final double? itemWidth;

  /// to run the default animation.
  final Duration animationDuration;

  /// Create a Horizontal Featured List.
  /// This widget contains:
  /// - <Title> is created by [titleText] and [titleTextStyle] at the top-left.
  /// - <See All> text button is created by [seeAllText] and [seeAllTextStyle] at the top-right, it has an [onPressedSeeAll] action helps in navigating to a full listing screen.
  /// - <Featured List View> is created by [itemBuilder], [itemCount] at the bottom.
  ///
  /// This widget has a default design helps you in saving time to create many of same designs.
  /// - [margin] : this is the spacing from parent to the left side of <Title>, and from parent to the right side of <See All>, and spacing between <Featured List View> items also.
  /// - Each item in <Featured List View> has a default style with [itemPadding], [overlayColor] is used for pressing animation, [itemBorderRadius] is the border radius, [onPressedItem] to navigate to a detail screen, [itemWidth] & [itemHeight] lets the horizontal [ListView] knows its children size.
  /// - [itemColorBuilder] : is background color of an item.
  /// - [animationDuration] : to run the default animation :D, use [Duration.zero] if you don't want it.

  const HorizontalFeaturedList({
    Key? key,
    this.itemHeight = 120,
    this.titleTextStyle = const TextStyle(
      fontSize: 18,
      fontWeight: FontWeight.bold,
      color: Colors.black87,
    ),
    this.seeAllTextStyle: const TextStyle(
      fontSize: 14,
      fontWeight: FontWeight.w500,
      color: Colors.black54,
    ),
    required this.onPressedSeeAll,
    this.margin: 24,
    this.itemPadding: const EdgeInsets.all(12),
    this.overlayColor: const Color(0x1F000000),
    this.itemBorderRadius: const BorderRadius.all(Radius.circular(22)),
    required this.onPressedItem,
    required this.itemBuilder,
    required this.itemCount,
    this.itemColorBuilder,
    required this.titleText,
    required this.seeAllText,
    this.itemWidth,
    this.animationDuration = const Duration(milliseconds: 400),
  }) : super(key: key);
  @override
  _HorizontalFeaturedListState createState() => _HorizontalFeaturedListState();
}

class _HorizontalFeaturedListState extends State<HorizontalFeaturedList>
    with TickerProviderStateMixin {
  late final AnimationController _animationController;

  @override
  void initState() {
    _animationController =
        AnimationController(vsync: this, duration: widget.animationDuration);
    _animationController.addListener(() {
      setState(() {});
    });
    _animationController.forward();
    super.initState();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [
        Padding(
          padding: EdgeInsets.only(
              left: widget.margin, right: max(0.0, widget.margin - 12)),
          child: Row(
            children: [
              Expanded(
                child: Text(
                  widget.titleText,
                  style: widget.titleTextStyle,
                ),
              ),
              ElevatedButton(
                child: Text(
                  widget.seeAllText,
                  style: widget.seeAllTextStyle,
                ),
                onPressed: widget.onPressedSeeAll,
                style: ButtonStyle(
                  shape: MaterialStateProperty.all<OutlinedBorder>(
                      RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(8))),
                  padding:
                      MaterialStateProperty.all<EdgeInsets>(EdgeInsets.all(12)),
                  elevation: MaterialStateProperty.all<double>(0.0),
                  tapTargetSize: MaterialTapTargetSize.shrinkWrap,
                  backgroundColor:
                      MaterialStateProperty.all<Color>(Colors.transparent),
                  overlayColor:
                      MaterialStateProperty.all<Color>(widget.overlayColor),
                  minimumSize: MaterialStateProperty.all<Size>(Size.zero),
                ),
              ),
            ],
          ),
        ),
        const SizedBox(height: 4),
        Container(
          height: widget.itemHeight,
          child: ListView.builder(
              scrollDirection: Axis.horizontal,
              itemCount: widget.itemCount,
              itemBuilder: (context, index) {
                final itemWidth = widget.itemWidth ??
                    MediaQuery.of(context).size.width * 0.75;
                final begin = Offset(-index * 0.95, 0);
                final firstWeight = 0.5;
                return SlideTransition(
                  position: TweenSequence<Offset>([
                    TweenSequenceItem<Offset>(
                        tween: Tween<Offset>(
                          begin: begin,
                          end: begin,
                        ),
                        weight: firstWeight),
                    TweenSequenceItem<Offset>(
                        tween: Tween<Offset>(
                          begin: begin,
                          end: Offset.zero,
                        ).chain(CurveTween(curve: Curves.linearToEaseOut)),
                        weight: 1 - firstWeight),
                  ]).animate(_animationController),
                  child: Container(
                    width: itemWidth,
                    margin: EdgeInsets.only(
                        left: index == 0 ? widget.margin : 0,
                        right: widget.margin),
                    child: ElevatedButton(
                      child: widget.itemBuilder(context, index),
                      onPressed: widget.onPressedItem,
                      style: ButtonStyle(
                        shape: MaterialStateProperty.all<OutlinedBorder>(
                            RoundedRectangleBorder(
                                borderRadius: widget.itemBorderRadius)),
                        padding: MaterialStateProperty.all<EdgeInsets>(
                            widget.itemPadding),
                        elevation: MaterialStateProperty.all<double>(0.0),
                        tapTargetSize: MaterialTapTargetSize.shrinkWrap,
                        backgroundColor: MaterialStateProperty.all<Color>(
                            widget.itemColorBuilder != null
                                ? widget.itemColorBuilder!(context, index)
                                : Colors.transparent),
                        overlayColor: MaterialStateProperty.all<Color>(
                            widget.overlayColor),
                        minimumSize: MaterialStateProperty.all<Size>(Size.zero),
                      ),
                    ),
                  ),
                );
              }),
        ),
      ],
    );
  }
}


example/lib/main.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:flutter_horizontal_featured_list/flutter_horizontal_featured_list.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  List<Color> colors = List.generate(20, (index) => randomColor());

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Example'),
        ),
        backgroundColor: Colors.white,
        body: Center(
          child: HorizontalFeaturedList(
            itemColorBuilder: (context, index) => colors[index],
            itemCount: colors.length,
            itemBuilder: (BuildContext context, int index) {
              return Column(
                children: [
                  Row(
                    children: <Widget>[
                      Icon(
                        Icons.face,
                        size: 24,
                        color: Colors.black45,
                      ),
                      Expanded(
                        child: Padding(
                          padding: const EdgeInsets.only(left: 8.0),
                          child: Text(
                            '($index) What is Lorem Ipsum?',
                            style: TextStyle(
                              fontSize: 16,
                              fontWeight: FontWeight.w500,
                            ),
                          ),
                        ),
                      ),
                    ],
                  ),
                  Expanded(
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Text(
                        '($index) Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
                        style: TextStyle(
                          fontSize: 12,
                        ),
                      ),
                    ),
                  ),
                ],
              );
            },
            onPressedItem: () {},
            onPressedSeeAll: () {},
            titleText: 'Ongoing Projects',
            seeAllText: 'See All',
          ),
        ),
      ),
    );
  }
}

Color randomColor() {
  var g = math.Random.secure().nextInt(255);
  var b = math.Random.secure().nextInt(255);
  var r = math.Random.secure().nextInt(255);
  return Color.fromARGB(255, r, g, b);
}
